docs/src/proposals/interest-bearing-tokens.md

```diff
@@ -0,0 +1,284 @@
+of this, we have to compound interest before any minting, burning, and
+transferring.
+
+### Initial Attempt
+
+The concept seems simple to implement. We can create an SPL token, allow
+tokens to be minted by a new "interest accruing" program, and then make sure to
+call it regularly.
+
+This solution has a few problems. First, how do we compound interest regularly?
+We can create a separate service which periodically sends transactions to update
+all of the accounts. Unfortunately, this will cost the service a lot, especially if
+there are thousands of accounts to update.
+
+We can enforce compounding before any transfer / mint / burn, but then all programs
+and web consumers need to change to make these update calls. How do we change programs
+and wallets to properly update accounts without creating a big `if is_i_token` case
+all over the code?
+
+This solution creates friction for almost all SPL token users: on-chain programs,
+wallets, applications, and end-users.
+
+## Proposed Solution
+
+Following the approach taken by Aave, we can lean on a crucial feature of i-tokens:
+although the token holding amounts are constantly changing, each token holder's
+proportion of total supply stays the same. Therefore, we can satisfy the
+fungibility requirement for tokens on proportional ownership. The actual token
+amounts used for token transfers / mints / burns are just an interface, and
+everything is converted to proportional shares inside the program.
+
+Let's go through an example. Imagine we have an i-token mint with a total of
+10,000 proportional shares, a 20% annual interest rate, and 1 share is currently
+worth 100 tokens. If I transfer 1,000 tokens today, the program converts that and moves
+10 shares. One year later, 1 share is worth 120 tokens, so if I transfer
+1,000 tokens, the program converts and only moves 8.33 shares. The same concept
+applies to mints and burns.
+
+Following this model, we never have to worry about compounding interest. We
+simply need to call the right program to perform instructions.
+
+## Required Changes
+
+A new interest-bearing token program is only one component of the overall solution,
+since i-tokens need to be easily integrated in on-chain programs, web apps,
+and wallets.
+
+### Token Program Conformance
+
+Since our solution entails the creation of a new token program, we're opening the
+floodgates for other app developers to do the same. To ensure safe and easy integration
+of new token programs, we need a set of conformance tests based on the current
+SPL token test suite.
+
+A new token program must include all of the same instructions as SPL token,
+but the "unchecked" variants should return errors. Concretely, this means the
+program must implement:
+
+* `InitializeMint`
+* `InitializeAccount`
+* `InitializeMultisig` -- TODO deprecate this and use a more general multisig?
+* `Revoke`
+* `SetAuthority`
+* `CloseAccount`
+* `FreezeAccount`
+* `ThawAccount`
+* `TransferChecked`
```

Comment:
This new `TransferChecked` accepts amount in terms of token amount from `GetHoldingBalance`, not in terms of share amount?

If the amount is continuously increasing indeed, I think we need some magic number recognition (`-1`) to move all of amount out of the given holding account to `CloseAccount`.

Reply:
Yes, the idea is that all interactions from the outside are in terms of tokens, and not shares.

That's a good point regarding draining everything out of an account.  Are you suggesting adding an optional holding account parameter to `CloseAccount` which moves whatever is left in the holding to another account?

Reply:
yeah, some thing like that? If share's amount is changing at slot generality, i thought it might hard to transfer all of then at once by end users.

I like `TransferChecked(u64::max)` than `Close()` because it's more like `transfer` than `close`. at user side, such ALL transfer can always paired with `Close()` instruction.

Reply:
> I like `TransferChecked(u64::max)` than `Close()` because it's more like `transfer` than `close`. at user side, such ALL transfer can always paired with `Close()` instruction.

I'm a little weary of setting the precedent of using sentinels like this, especially in an amount field.  A `TransferAllChecked(mint)` instruction would work just as well and share most logic with `TransferChecked(amount,mint)`

Reply:
Yeah, I like `TransferAllChecked(mint)` :)

Reply:
@joncinque Could you add this new `TransferCheckedAll` or `TransferAllChecked` into the text? so that we can mark this thread as resolved?

Reply:
Yes, thank you for your feedback!